source/simulator/helper_functions.cc

```diff
@@ -1574,6 +1574,88 @@ namespace aspect
                                "Please check the consistency of your input file."));
       }
   }
+
+
+
+  template <int dim>
+  double
+  Simulator<dim>::compute_initial_newton_residual(LinearAlgebra::BlockVector &linearized_stokes_initial_guess)
+  {
+    LinearAlgebra::BlockVector temp_linearization_point = current_linearization_point;
+    const unsigned int block_vel = introspection.block_indices.velocities;
+    temp_linearization_point.block(introspection.block_indices.velocities) = 0;
+    linearized_stokes_initial_guess.block (block_vel) = 0;
+
+    rebuild_stokes_matrix = assemble_newton_stokes_system = assemble_newton_stokes_matrix = true;
+    rebuild_stokes_preconditioner = false;
+
+    denormalize_pressure (this->last_pressure_normalization_adjustment,
+                          linearized_stokes_initial_guess,
+                          temp_linearization_point);
+
+    compute_current_constraints ();
+    assemble_stokes_system();
+
+    this->last_pressure_normalization_adjustment = normalize_pressure(temp_linearization_point);
+
+    const double initial_newton_residual_velo = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+    const double initial_newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+    const double initial_newton_residual = std::sqrt(initial_newton_residual_velo * initial_newton_residual_velo + initial_newton_residual_pres * initial_newton_residual_pres);
+
+    pcout << "   Initial Newton Stokes residual = " << initial_newton_residual << ", v = " << initial_newton_residual_velo << ", p = " << initial_newton_residual_pres << std::endl << std::endl;
+    return initial_newton_residual;
+  }
+
+
+
+  template <int dim>
+  void
+  Simulator<dim>::compute_Eisenstat_Walker_linear_tolerance(const bool EisenstatWalkerChoiceOne,
```

Comment:
I would prefer if you could make the `linear_stokes_solver_tolerance` parameter a `const double` and return the newly computed value. The call to this function could then look like
```
linear_solver_tolerance = compute_Eisenstat_Walker_linear_tolerance( ..., linear_solver_tolerance, ...);
```
It is good practice to avoid functions that modify their parameters, where possible (it is not always possible). They make the code harder to understand.